oak/server/enclave_server.h

```diff
@@ -139,61 +141,33 @@ class EnclaveServer final : public asylo::TrustedApplication {
     }
   }
 
-  // Creates a new gRPC event tag based on a number. gRPC tags are usually employed to uniquely
-  // identify events, but we do not need to keep track of them, and therefore we just create them
-  // based on arbitrary numbers.
-  // See https://grpc.io/grpc/cpp/classgrpc_1_1_completion_queue.html.
-  static void *tag(int i) { return reinterpret_cast<void *>(i); }
-
-  // Processes the next available event from the completion queue, blocking if none is available.
-  void ProcessNextEvent() {
-    void *out_tag;
-    bool ok = false;
-    completion_queue_->Next(&out_tag, &ok);
-    // TODO: Return the value of ok so that we can break the completion queue loop.
-    if (!ok) {
-      LOG(INFO) << "Could not process gRPC event from completion queue";
-    }
+  void RequestNextCall() {
+    // The stream will delete itself when it finishes.
+    auto stream = std::make_shared<grpc_server::GrpcStream>(node_);
+    generic_service_.RequestCall(&stream->server_context(), &stream->server_reader_writer(),
+                                 completion_queue_.get(), completion_queue_.get(),
+                                 new grpc_server::StreamCreationEvent(stream));
   }
 
   // Consumes gRPC events from the completion queue in an infinite loop.
   void CompletionQueueLoop() {
     LOG(INFO) << "Starting gRPC completion queue loop";
-    int i = 0;
+    RequestNextCall();
     while (true) {
-      ::grpc::GenericServerContext context;
-      ::grpc::GenericServerAsyncReaderWriter stream(&context);
-
-      // We request a new event corresponding to a generic call. This will create an entry in the
-      // completion queue when a new call is available.
-      generic_service_.RequestCall(&context, &stream, completion_queue_.get(),
-                                   completion_queue_.get(), tag(i++));
-
-      // Wait for a generic call to arrive.
-      ProcessNextEvent();
-
-      // Read request data.
-      ::grpc::ByteBuffer request;
-      stream.Read(&request, tag(i++));
-
-      // Process the event related to the read we just requested.
-      ProcessNextEvent();
-
-      ::grpc::ByteBuffer response;
-
-      // Invoke the actual gRPC handler on the Oak Node.
-      ::grpc::Status status = node_->HandleGrpcCall(&context, &request, &response);
-      if (!status.ok()) {
-        LOG(INFO) << "Failed: " << status.error_message();
+      grpc_server::BaseGrpcEvent *tag = nullptr;
+      bool ok = false;
+      if (!completion_queue_->Next(reinterpret_cast<void **>(&tag), &ok)) {
+        return;
       }
 
-      ::grpc::WriteOptions options;
-
-      // Write response data.
-      stream.WriteAndFinish(response, options, status, tag(i++));
+      if (ok) {
```

Comment:
Log something / early return if false? In fact I think we would always want to treat this as if `Next` itself returned false, right? So maybe you could chain the two conditions together above.

Reply:
I change this and added a log